raiden/transfer/mediated_transfer/initiator_manager.py

```diff
@@ -64,129 +68,183 @@ def events_for_cancel_current_route(transfer_description) -> List[Event]:
-        assert payment_state.initiator is None, msg
 
+def handle_init(
+        payment_state: InitiatorPaymentState,
+        state_change: ActionInitInitiator,
+        channelidentifiers_to_channels: ChannelMap,
+        pseudo_random_generator: random.Random,
+        block_number: BlockNumber,
+) -> TransitionResult:
+    events: List[Event]
+    if payment_state is None:
         sub_iteration = initiator.try_new_route(
-            old_initiator_state=old_initiator_state,
+            old_initiator_state=None,
             channelidentifiers_to_channels=channelidentifiers_to_channels,
             available_routes=state_change.routes,
-            transfer_description=transfer_description,
+            transfer_description=state_change.transfer,
             pseudo_random_generator=pseudo_random_generator,
             block_number=block_number,
         )
 
-        events.extend(cancel_events)
-        events.extend(sub_iteration.events)
-        assert sub_iteration.new_state
-        payment_state.initiator = sub_iteration.new_state
+        events = sub_iteration.events
+        if sub_iteration.new_state:
+            payment_state = InitiatorPaymentState(
+                initiator_transfers={
+                    sub_iteration.new_state.transfer.lock.secrethash: sub_iteration.new_state,
+                },
+            )
+    else:
+        events = list()
 
     iteration = TransitionResult(payment_state, events)
-
     return iteration
 
 
 def handle_cancelpayment(
         payment_state: InitiatorPaymentState,
-        channel_state: NettingChannelState,
+        channelidentifiers_to_channels: ChannelMap,
 ) -> TransitionResult:
     """ Cancel the payment and all related transfers. """
     # Cannot cancel a transfer after the secret is revealed
-    if can_cancel(payment_state):
-        transfer_description = payment_state.initiator.transfer_description
-        cancel_events = cancel_current_route(payment_state)
-
-        cancel = EventPaymentSentFailed(
-            payment_network_identifier=channel_state.payment_network_identifier,
-            token_network_identifier=channel_state.token_network_identifier,
-            identifier=transfer_description.payment_identifier,
-            target=transfer_description.target,
-            reason='user canceled payment',
-        )
-        cancel_events.append(cancel)
+    events = list()
+    for secrethash in list(payment_state.initiator_transfers.keys()):
+        initiator_state = payment_state.initiator_transfers[secrethash]
+        channel_identifier = initiator_state.channel_identifier
+        channel_state = channelidentifiers_to_channels[channel_identifier]
+        if can_cancel(initiator_state):
+            transfer_description = initiator_state.transfer_description
+            cancel_events = cancel_current_route(payment_state, initiator_state)
+
+            del payment_state.initiator_transfers[secrethash]
```

Comment:
we must *not* remove the initiator task from the state, otherwise there will be nothing to remove the expired lock.

Reply:
You're right. I'll amend.